File: personal/investments/scripts/src/mask.ts
// Mask real account codes, redact personal names, and detect account kinds.
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";

const CODE_RE = /-([A-Za-z0-9]+)\.csv$/;
const STMT_RE = /-\d{4}-\d{2}-\d{2}-monthly-statement.*$/;

const KIND_RULES: ReadonlyArray<readonly [string, string]> = [
  ["credit-card", "CreditCard"],
  ["managed (tfsa)", "ManagedTFSA"],
  ["direct indexing", "DirectIndexing"],
  ["family-resp", "RESP"],
  ["resp", "RESP"],
  ["rrsp", "RRSP"],
  ["home", "FHSA"],
  ["fhsa", "FHSA"],
  ["tfsa", "TFSA"],
  ["non-registered", "NonRegistered"],
  ["chequing", "Chequing"],
  ["savings", "Savings"],
  ["us dollars", "USD"],
  ["us_dollars", "USD"],
  ["crypto", "Crypto"],
];

export interface Redactions {
  readonly names: string[];
  readonly accountLabelPeople: Record<string, string>;
}

function sha256Hex(value: string): string {
  return createHash("sha256").update(value, "utf-8").digest("hex");
}

export function maskAccountCode(realCode: string): string {
  return `acct_${sha256Hex(realCode).slice(0, 8)}`;
}

export function shortAccountId(realCode: string): string {
  return sha256Hex(realCode).slice(0, 4);
}

export function accountCodeFromFilename(name: string): string {
  if (name.includes("credit-card")) return "card";
  const match = CODE_RE.exec(name);
  if (!match?.[1]) throw new Error(`No account code found in filename: ${name}`);
  return match[1];
}

export function accountNameFromFilename(name: string): string {
  if (name.includes("credit-card")) return "Card";
  const stem = name
    .replace(STMT_RE, "")
    .replace(/^[^\w(]+/, "")
    .trim();
  return stem || "Account";
}

export function detectKind(filename: string): string {
  const lowered = filename.toLowerCase();
  for (const [needle, kind] of KIND_RULES) {
    if (lowered.includes(needle)) return kind;
  }
  if (lowered.startsWith("pe-")) return "PE";
  return "Other";
}

export function loadRedactions(path: string): Redactions {
  const data = JSON.parse(readFileSync(path, "utf-8")) as {
    names?: string[];
    account_label_people?: Record<string, string>;
  };
  return { names: data.names ?? [], accountLabelPeople: data.account_label_people ?? {} };
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

export function redact(text: string, red: Redactions): string {
  let result = text;
  const ordered = [...red.names].sort((a, b) => b.length - a.length);
  for (const name of ordered) {
    result = result.replace(new RegExp(escapeRegExp(name), "gi"), "[redacted]");
  }
  return result;
}

File: personal/investments/scripts/investments/mask.py
"""Mask real account codes, redact personal names, and detect account kinds."""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass
from pathlib import Path

_CODE_RE = re.compile(r"-([A-Za-z0-9]+)\.csv$")
_KIND_RULES: list[tuple[str, str]] = [
    ("credit-card", "CreditCard"),
    ("managed (tfsa)", "ManagedTFSA"),
    ("direct indexing", "DirectIndexing"),
    ("family-resp", "RESP"),
    ("resp", "RESP"),
    ("rrsp", "RRSP"),
    ("home", "FHSA"),
    ("fhsa", "FHSA"),
    ("tfsa", "TFSA"),
    ("non-registered", "NonRegistered"),
    ("chequing", "Chequing"),
    ("savings", "Savings"),
    ("us dollars", "USD"),
    ("us_dollars", "USD"),
    ("crypto", "Crypto"),
]


@dataclass(frozen=True)
class Redactions:
    """Configured personal-name redaction data."""

    names: list[str]
    account_label_people: dict[str, str]


def mask_account_code(real_code: str) -> str:
    """Return a stable masked id for a real account code; never store the real code."""
    digest = hashlib.sha256(real_code.encode("utf-8")).hexdigest()
    return f"acct_{digest[:8]}"


def account_code_from_filename(name: str) -> str:
    """Return the trailing real account code, or 'card' for credit-card files."""
    if "credit-card" in name:
        return "card"
    match = _CODE_RE.search(name)
    if not match:
        raise ValueError(f"No account code found in filename: {name}")
    return match.group(1)


def short_account_id(real_code: str) -> str:
    """Return a short 4-hex disambiguator for a real code; the real code is never stored."""
    return hashlib.sha256(real_code.encode("utf-8")).hexdigest()[:4]


_STMT_RE = re.compile(r"-\d{4}-\d{2}-\d{2}-monthly-statement.*$")


def account_name_from_filename(name: str) -> str:
    """Return the human account label from a filename (the part before the date)."""
    if "credit-card" in name:
        return "Card"
    stem = _STMT_RE.sub("", name)
    stem = re.sub(r"^[^\w(]+", "", stem).strip()
    return stem or "Account"


def detect_kind(filename: str) -> str:
    """Detect the account kind from a filename using ordered keyword rules."""
    lowered = filename.lower()
    for needle, kind in _KIND_RULES:
        if needle in lowered:
            return kind
    if lowered.startswith("pe-"):
        return "PE"
    return "Other"


def load_redactions(path: Path) -> Redactions:
    """Load the redaction config from JSON."""
    data = json.loads(path.read_text(encoding="utf-8"))
    return Redactions(
        names=list(data.get("names", [])),
        account_label_people=dict(data.get("account_label_people", {})),
    )


def redact(text: str, red: Redactions) -> str:
    """Replace configured personal names in text with a redaction marker."""
    result = text
    for name in sorted(red.names, key=len, reverse=True):
        result = re.sub(re.escape(name), "[redacted]", result, flags=re.IGNORECASE)
    return result
